File: 15.BinarySearch二分查找/leetcode_3488_solveQueries.py
from typing import List
from collections import defaultdict
from bisect import *

class Solution:
    def solveQueries(self, nums: List[int], queries: List[int]) -> List[int]:
        n = len(nums)
        m = len(queries)
        nums = nums+nums
        pos = defaultdict(list)
        for i in range(len(nums)):
            pos[nums[i]].append(i)
        for i in range(m):
            queries.append(queries[i]+n)
        
        res = [10**18]*m
        for i in range(len(queries)):
            x = nums[queries[i]]
            indx = bisect_left(pos[x], queries[i])
            if indx > 0:
                res[i%m] = min(res[i%m], queries[i]-pos[x][indx-1])
            if indx+1 < len(pos[x]):
                res[i%m] = min(res[i%m], pos[x][indx+1]-queries[i])
        return [(-1 if i==n else i) for i in res]

if __name__ == "__main__":
    s = Solution()
    nums = [1,3,1,4,1,3,2]
    queries = [0,3,5]
    print(s.solveQueries(nums, queries))

File: 15.BinarySearch二分查找/leetcode_3488_solveQueries.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> solveQueries(vector<int>& nums, vector<int>& queries) {
        unordered_map<int, vector<int>> pos;
        int n = nums.size();
        for (int i=0; i<n; i++){
            pos[nums[i]].push_back(i-n);
            pos[nums[i]].push_back(i);
            pos[nums[i]].push_back(i+n);
        }
        for (auto& kv : pos){
            sort(pos[kv.first].begin(), pos[kv.first].end());
        }
        vector<int> res(queries.size(), -1);
        for (int i=0; i<queries.size(); i++){
            int cur = queries[i], x = nums[cur];
            if (pos[x].size() == 3){
                continue;
            }
            int indx = lower_bound(pos[x].begin(), pos[x].end(), cur)-pos[x].begin();
            res[i] = min(cur-pos[x][indx-1], pos[x][indx+1]-cur);
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> nums = {1,3,1,4,1,3,2};
    vector<int> queries = {0,3,5};
    vector<int> res = s.solveQueries(nums, queries);
    for (int i : res){
        cout << i << " ";
    }
    cout << endl;
}
